presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java

```diff
@@ -23,43 +23,39 @@
 import com.facebook.presto.sql.planner.plan.PlanNode;
 
 import java.util.Optional;
+import java.util.function.Predicate;
+
+import static com.facebook.presto.util.Optionals.tryCast;
+import static com.facebook.presto.util.Predicates.predicate;
 
 public class MergeLimitWithDistinct
-    implements Rule
+        implements Rule
 {
+    private static final Predicate<AggregationNode> NOT_DISTINCT = predicate(MergeLimitWithDistinct::isDistinct).negate();
+
     @Override
     public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
     {
-        if (!(node instanceof LimitNode)) {
-            return Optional.empty();
-        }
-
-        LimitNode parent = (LimitNode) node;
-
-        PlanNode input = lookup.resolve(parent.getSource());
-        if (!(input instanceof AggregationNode)) {
-            return Optional.empty();
-        }
-
-        AggregationNode child = (AggregationNode) input;
+        Optional<LimitNode> parent = tryCast(node, LimitNode.class);
 
-        if (isDistinct(child)) {
-            return Optional.empty();
-        }
+        Optional<AggregationNode> childNotDistinct = parent
+                .flatMap(p -> lookup.resolve(p.getSource(), AggregationNode.class))
+                .filter(NOT_DISTINCT);
 
-        return Optional.of(
-                new DistinctLimitNode(
-                        parent.getId(),
-                        child.getSource(),
-                        parent.getCount(),
-                        false,
-                        child.getHashSymbol()));
+        return parent.flatMap(p ->
+                childNotDistinct.map(d ->
+                        new DistinctLimitNode(
+                                p.getId(),
+                                lookup.resolve(d.getSource()),
+                                p.getCount(),
+                                false,
+                                d.getHashSymbol())));
     }
 
     /**
      * Whether this node corresponds to a DISTINCT operation in SQL
      */
-    private boolean isDistinct(AggregationNode node)
+    private static boolean isDistinct(AggregationNode node)
```

Comment:
I think you have a point here, but this is not the scope of this PR. A quick search revealed 3 occurrences of similar code throughout the codebase. Feel free to open an issue for unifying those.